### PR TITLE
feat(agent/agentssh): use tcp for X11 forwarding

### LINKED ISSUE #14198: X11 forwarding should use network instead of unix socket and support being forwarded multiple times
Our current X11 forwarding implementation (in `agentssh`) uses Unix socket forwarding and places the socket file in `/tmp/.X11-unix/X0`.

New SSH connections overwrite the socket and take over communication. This means that connecting with `ssh -X` twice, then disconnecting from the newest session will also disconnect the old session.

OpenSSH typically does X11 forwarding over port `6000` and up, and generally starting from an offset of `10` (i.e. `6010`). For OpenSSH the `DISPLAY` env looks commonly like: `DISPLAY=localhost:10.0` (referring to port `6010`), whereas `agentssh` always sets this to `DISPLAY=:0.0` (referring to `/tmp/.X11-unix/X0`).

The `agentssh` behavior is problematic in a few situations:

1. The agent doesn't have permission to write to `/tmp/.X11-unix`
2. There's an X server running inside the workspace (will have allocated `/tmp/.X11-unix/X0` already)
3. The user wants the X11 connection to remain open even if other SSH sessions are closed

I propose we implement the listening port + start offset to avoid the conflict listed in 2.

An alternative method is to try `/tmp/.X11-unix/X[0-9]+` until we find the next free slot and set the `DISPLAY` environment variable accordingly. (And if we use an offset it's unlikely to collide with X running in the workspace.)

Please note that `ssh.X11.ScreenNumber` in `x11Callback` should _not_ decide the number we use in the workspace, it should be the next free port or file.

### PR DESCRIPTION
**Pull request by mafredri**
This PR changes X11 forwarding from socket-based to TCP based. This matches OpenSSH behavior where the _client_ forwards from `/tmp/.X11-unix/X*` and the server listens on `localhost:6000(+)`.

Since we can now have multiple SSH connections that each handle their individual X11 forwarding, we can end up adding more non-duplicate entries into `~/.Xauthority` as well. Ideally we would clean our entries up when the connection closes, but that's more of a nice-to-have and not part of this PR.

Note that X11 forwarding is still only supported by OpenSSH clients (`coder config-ssh`), this PR did not add X11 forwarding support to `coder ssh`.

Fixes #14198

### CODE REVIEW

File: agent/agentssh/x11.go [hunk 1]
```diff
@@ -22,102 +23,133 @@ import (
...
 	if err != nil {
+		s.logger.Warn(ctx, "failed to get hostname", slog.Error(err))
+		s.metrics.x11HandlerErrors.WithLabelValues("hostname").Add(1)
+		return -1, false
+	}
+
+	var (
+		lc   net.ListenConfig
+		ln   net.Listener
+		port = X11StartPort + *s.config.X11DisplayOffset
```
Comment (johnstcn) [inline agent/agentssh/x11.go:59]: Should we check if `X11DisplayOffset != nil`?

File: agent/agentssh/x11.go [hunk 2]
```diff
@@ -22,102 +23,133 @@ import (
...
+			display = port - X11StartPort
+			break
+		}
+	}
+	if ln == nil {
 		s.logger.Warn(ctx, "failed to listen for X11", slog.Error(err))
-		return false
+		s.metrics.x11HandlerErrors.WithLabelValues("listen").Add(1)
+		return -1, false
+	}
```
Comment (johnstcn) [inline agent/agentssh/x11.go:73]: Could we extract this to its own helper function?

File: agent/agentssh/x11.go [hunk 3]
```diff
@@ -22,102 +23,133 @@ import (
...
 	}
-	s.trackListener(listener, true)
 
 	go func() {
-		defer listener.Close()
-		defer s.trackListener(listener, false)
-		handledFirstConnection := false
+		// Don't leave the listener open after the session is gone.
+		<-ctx.Done()
+		_ = ln.Close()
```
Comment (johnstcn) [inline agent/agentssh/x11.go:92]: Should we put this in a `sync.Once`?

File: agent/agentssh/x11.go [hunk 4]
```diff
@@ -22,102 +23,133 @@ import (
...
+				continue
 			}
-			unixAddr, ok := unixConn.LocalAddr().(*net.UnixAddr)
+			tcpAddr, ok := tcpConn.LocalAddr().(*net.TCPAddr)
 			if !ok {
-				s.logger.Warn(ctx, fmt.Sprintf("failed to cast local address to UnixAddr. got: %T", unixConn.LocalAddr()))
-				return
+				s.logger.Warn(ctx, fmt.Sprintf("failed to cast local address to TCPAddr. got: %T", tcpConn.LocalAddr()))
+				_ = conn.Close()
+				continue
```
Comment (johnstcn) [inline agent/agentssh/x11.go:123]: If I understand this correctly, the net effect here could be X11 forwarding silently breaking under certain conditions. Should this be a more explicit error?

File: agent/agentssh/x11.go [hunk 5]
```diff
@@ -22,102 +23,133 @@ import (
...
-	// We want to overwrite the socket so that subsequent connections will succeed.
-	socketPath := filepath.Join(s.config.X11SocketDir, fmt.Sprintf("X%d", x11.ScreenNumber))
-	err := os.Remove(socketPath)
-	if err != nil && !errors.Is(err, os.ErrNotExist) {
-		s.logger.Warn(ctx, "failed to remove existing X11 socket", slog.Error(err))
-		return false
-	}
-	listener, err := net.Listen("unix", socketPath)
+
+	hostname, err := os.Hostname()
```
Comment (johnstcn) [inline agent/agentssh/x11.go:49]: Does this actually need to be the 'real' hostname? Could we just make one up based on the workspace+agent names?

Comment (mafredri) [reply agent/agentssh/x11.go:49]: Yeah, hostname is required for `FamilyLocal` authentication.

File: agent/agentssh/x11.go (hunk 1 above)
Comment (mafredri) [reply agent/agentssh/x11.go:59]: It's guaranteed to be set (as per docs has a default value), if it's not the server is being created without `NewServer` which is not supported.

The test should catch it if this is ever not the case 👍🏻

File: agent/agentssh/x11.go (hunk 3 above)
Comment (mafredri) [reply agent/agentssh/x11.go:92]: Nah, there's no reason to protect it from multiple close.

File: agent/agentssh/x11.go (hunk 4 above)
Comment (mafredri) [reply agent/agentssh/x11.go:123]: In theory this can't happen, except on maybe some esoteric system? What would you like to see done differently?

Comment (johnstcn) [reply agent/agentssh/x11.go:111]: Fair enough 👍  I'm not sure what would be better here. This is happening in the agent itself if I understand correctly, so a failure here would just end up in the agent logs. To troubleshoot this you would need to look in the agent logs in any case. I don't have a better suggestion for this theoretical possibility, so let's leave it for now!

**Review by johnstcn — APPROVED** [review]
Nothing else blocking on my end!

**Pull request merged**